08_irq/poll_button/khmod.c

```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: MIT and GPL
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/gpio.h>
+#include <linux/timer.h>
+
+MODULE_DESCRIPTION("Button poll module");
+MODULE_AUTHOR("vadym.kharchuk@globallogic.com");
+MODULE_VERSION("0.1");
+MODULE_LICENSE("Dual MIT/GPL");
+
+#define POLL_TIME_MS	20
+#define BLINK_TIME_MS	1
+
+#define GPIO_NUMBER(port, bit) (32 * (port) + (bit))
+
+#define LED_SD  GPIO_NUMBER(1, 22)
+#define LED_MMC GPIO_NUMBER(1, 24)
+
+/* On-board button.
+ *
+ * LCD/HDMI interface must be disabled.
+ */
+#define BUTTON  GPIO_NUMBER(2, 8)
+
+static int button_gpio = -1;
+
+struct timer_list tim_l;
+struct hrtimer tim_h;
+
+enum hrtimer_restart hr_callback(struct hrtimer *timer)
+{
+	gpio_set_value(LED_MMC, 0);
+	return HRTIMER_NORESTART;
+}
+
+static void button_poll(struct timer_list *data)
+{
+	static bool old_button_state;
+	bool button_state;
+	ktime_t kt;
+
+	button_state = gpio_get_value(BUTTON);
+
+	if (old_button_state != button_state) {
+		gpio_set_value(LED_MMC, 1);
+		kt = ktime_set(0, BLINK_TIME_MS * 1000000);
+		hrtimer_start(&tim_h, kt, HRTIMER_MODE_REL);
+	} else if (!button_state) {
+		gpio_set_value(LED_SD, 1);
+	} else {
+		gpio_set_value(LED_SD, 0);
+	}
+
+	old_button_state = button_state;
+	mod_timer(&tim_l, jiffies + msecs_to_jiffies(POLL_TIME_MS));
+}
+
+static int led_gpio_init(int gpio)
+{
+	int rc;
+
+	rc = gpio_direction_output(gpio, 0);
+	if (rc)
+		return rc;
+
+	return 0;
+}
+
+static int button_gpio_init(int gpio)
+{
+	int rc;
+
+	rc = gpio_request(gpio, "Onboard user button");
+	if (rc)
+		goto err_register;
+
+	rc = gpio_direction_input(gpio);
+	if (rc)
+		goto err_input;
+
+	button_gpio = gpio;
+	pr_info("Init GPIO%d OK\n", button_gpio);
+	return 0;
+
+err_input:
+	gpio_free(gpio);
+err_register:
+	return rc;
+}
+
+static void button_gpio_deinit(void)
+{
+	if (button_gpio >= 0) {
+		gpio_free(button_gpio);
+		pr_info("Deinit GPIO%d\n", button_gpio);
+	}
+}
+
+/* Module entry/exit points */
+static int __init onboard_io_init(void)
+{
+	int rc;
+
+	rc = button_gpio_init(BUTTON);
+	if (rc) {
+		pr_err("Can't set GPIO%d for button\n", BUTTON);
+		goto err_button;
+	}
+
+	rc = led_gpio_init(LED_SD);
+	if (rc) {
+		pr_err("Can't set GPIO%d for output\n", LED_SD);
+		goto err_led;
+	}
+
+	rc = led_gpio_init(LED_MMC);
+	if (rc) {
+		pr_err("Can't set GPIO%d for output\n", LED_MMC);
+		goto err_led;
+	}
+
+	timer_setup(&tim_l, button_poll, 0);
+	add_timer(&tim_l);
+
+	hrtimer_init(&tim_h, CLOCK_MONOTONIC, HRTIMER_MODE_REL);
+	tim_h.function = &hr_callback;
+
+	return 0;
+
+err_led:
+	button_gpio_deinit();
```

Comment:
you init two leds.
what if first - ok, second - fail ?

Reply:
Nothing? In led_gpio_init we just set gpio_direction_output as output.